File: nodegs_java/node.gs.client/src/nodegsclient/service/UserSecurityService.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package nodegsclient.service;

import java.util.ArrayList;
import nodegsclient.base.Binary;
import nodegsclient.base.CMessage;
import nodegsclient.base.HttpPostService;

/**
 *
 * @author GAIA
 */
public class UserSecurityService extends HttpPostService {
        public UserSecurityService()
        {
            SetServiceID(SERVICEID_USERSECURITY);
        }
        
        /**
         * 自选股服务ID
        */
        public final int SERVICEID_USERSECURITY = 1;

        /**
         * 添加类别方法ID
         */
        public final int FUNCTIONID_USERSECURITY_ADDCATEGORIES = 0;

        /**
         * 删除类别方法ID
         */
        public final int FUNCTIONID_USERSECURITY_DELETECATEGORIES = 1;

        /**
         * 获取所有类别信息方法ID
         */
        public final int FUNCTIONID_USERSECURITY_GETCATEGORIES = 2;

        /**
         * 更新类别信息方法ID
         */
        public final int FUNCTIONID_USERSECURITY_UPDATECATEGORIES = 3;

        /**
         * 添加证券方法ID
         */
        public final int FUNCTIONID_USERSECURITY_ADDSECURITIES = 4;

        /**
         * 删除证券方法ID
         */
        public final int FUNCTIONID_USERSECURITY_DELETESECURITIES = 5;

        /**
         * 更新类别排序方法ID
         */
        public final int FUNCTIONID_USERSECURITY_UPDTAEORDERNUM = 6;
                
        /**
         * 添加自选股类别
         * @param requestID 请求ID
         * @param category 类别
         * @return 状态
         */
        public int AddCategory(int requestID, UserSecurityCategory category)
        {
            ArrayList<UserSecurityCategory> categories = new ArrayList<UserSecurityCategory>();
            categories.add(category);
            int ret = Send(FUNCTIONID_USERSECURITY_ADDCATEGORIES, requestID, categories);
            categories.clear();
            return ret > 0 ? 1 : 0;
        }
        
        /**
         * 添加证券
         * @param requestID 请求ID
         * @param category 类别
         * @return  状态
         */
        public int AddSecurities(int requestID, UserSecurityCategory category)
        {
            ArrayList<UserSecurityCategory> categories = new ArrayList<UserSecurityCategory>();
            categories.add(category);
            int ret = Send(FUNCTIONID_USERSECURITY_ADDSECURITIES, requestID, categories);
            categories.clear();
            return ret > 0 ? 1 : 0;
        }
        
        /**
         * 删除类别
         * @param requestID 请求ID
         * @param category 类别
         * @return  状态
         */
        public int DeleteCategory(int requestID, UserSecurityCategory category)
        {
            ArrayList<UserSecurityCategory> categories = new ArrayList<UserSecurityCategory>();
            categories.add(category);
            int ret = Send(FUNCTIONID_USERSECURITY_DELETECATEGORIES, requestID, categories);
            categories.clear();
            return ret > 0 ? 1 : 0;
        }
        
        /**
         * 删除证券
         * @param requestID 请求ID
         * @param category 类别
         * @return  状态
         */
        public int DeleteSecurities(int requestID, UserSecurityCategory category)
        {
            ArrayList<UserSecurityCategory> categories = new ArrayList<UserSecurityCategory>();
            categories.add(category);
            int ret = Send(FUNCTIONID_USERSECURITY_DELETESECURITIES, requestID, categories);
            categories.clear();
            return ret > 0 ? 1 : 0;
        }
        
        /**
         * 获取类别信息
         * @param requestID 请求ID
         * @param category 类别
         * @return  状态
         */
        public int GetCategories(int requestID, UserSecurityCategory category)
        {
            ArrayList<UserSecurityCategory> categories = new ArrayList<UserSecurityCategory>();
            categories.add(category);
            int ret = Send(FUNCTIONID_USERSECURITY_GETCATEGORIES, requestID, categories);
            categories.clear();
            return ret > 0 ? 1 : 0;
        }

        /**
         * 通过流获取类别信息列表
         * @param categories 类别信息
         * @param body 包体
         * @param bodyLength 包体长度
         * @return  状态
         */
        public static int GetCategories(ArrayList<UserSecurityCategory> categories, byte[] body, int bodyLength)
        {
            try
            {
                Binary br = new Binary();
                br.Write(body, bodyLength);
                int size = br.ReadInt();
                if (size > 0)
                {
                    for (int i = 0; i < size; i++)
                    {
                        UserSecurityCategory category = new UserSecurityCategory();
                        category.m_categoryID = br.ReadString();
                        category.m_userID = br.ReadInt();
                        category.m_name = br.ReadString();
                        category.m_codes = br.ReadString();
                        category.m_type = br.ReadInt();
                        category.m_orderNum = br.ReadInt();
                        categories.add(category);
                    }
                }
                br.Close();
            }
            catch(Exception ex)
            {
            }
            return 1;
        }

        /**
         * 接收数据
         */
        @Override
        public void OnReceive(CMessage message)
        {
            super.OnReceive(message);
            SendToListener(message);
        }
        /**
         * 发送消息
         * @param functionID 方法ID
         * @param requestID 请求ID
         * @param categories 类别集合
         * @return 状态
         */
        public int Send(int functionID, int requestID, ArrayList<UserSecurityCategory> categories)
        {
            try
            {
                Binary bw = new Binary();
                int categoriesSize = categories.size();
                bw.WriteInt(categoriesSize);
                for (int i = 0; i < categoriesSize; i++)
                {
                    UserSecurityCategory category = categories.get(i);
                    bw.WriteString(category.m_categoryID);
                    bw.WriteInt(category.m_userID);
                    bw.WriteString(category.m_name);
                    bw.WriteString(category.m_codes);
                    bw.WriteInt(category.m_type);
                    bw.WriteInt(category.m_orderNum);
                }
                byte[] bytes = bw.GetBytes();
                int ret = Send(new CMessage(GetGroupID(), GetServiceID(), functionID, GetSessionID(), requestID, 0, 0, GetCompressType(), bytes.length, bytes));
                bw.Close();
                return ret;
            }
            catch(Exception ex)
            {
            }        
            return 0;
        }
        
        /**
         * 更新类别信息
         * @param requestID 请求ID
         * @param category 类别
         * @return 状态
         */
        public int UpdateCategory(int requestID, UserSecurityCategory category)
        {
            ArrayList<UserSecurityCategory> categories = new ArrayList<UserSecurityCategory>();
            categories.add(category);
            int ret = Send(FUNCTIONID_USERSECURITY_UPDATECATEGORIES, requestID, categories);
            categories.clear();
            return ret > 0 ? 1 : 0;
        }

        /// <summary>
        /// 更新指标的排序
        /// </summary>
        /// <param name="requestID">请求ID</param>
        /// <param name="userID">用户ID</param>
        /// <param name="ids">ID</param>
        /// <returns>状态</returns>
        /**
         * 更新指标的排序
         * @param requestID 请求ID
         * @param userID 用户ID
         * @param ids IDS
         * @return 状态
         */
        public int UpdateOrderNum(int requestID, int userID, ArrayList<String> ids)
        {
            String str = "";
            int idsSize = ids.size();
            for (int i = 0; i < idsSize; i++)
            {
                str += ids.get(i);
                if (i != idsSize - 1)
                {
                    str += ",";
                }
            }
            UserSecurityCategory category = new UserSecurityCategory();
            category.m_categoryID = str;
            category.m_userID = userID;
            ArrayList<UserSecurityCategory> categories = new ArrayList<UserSecurityCategory>();
            categories.add(category);
            int ret = Send(FUNCTIONID_USERSECURITY_UPDTAEORDERNUM, requestID, categories);
            categories.clear();
            return ret > 0 ? 1 : 0;
        }
}

File: nodegs/nodegsclient/nodegsclient/Service/UserSecurityService.cpp
#include "..\\stdafx.h"
#include "..\\CStr.h"
#include "UserSecurityService.h"
#include "Binary.h"

namespace OwLibCT
{
	UserSecurityCategory::UserSecurityCategory()
	{
		m_categoryID = L"";
		m_userID = 0;
		m_name = L"";
		m_codes = L"";
		m_type = 0;
		m_orderNum = 0;
	}

	///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	UserSecurityService::UserSecurityService()
	{
		m_socketID = 0;
		SetServiceID(SERVICEID_USERSECURITY);
	}

	UserSecurityService::~UserSecurityService()
	{
	}

	int UserSecurityService::GetSocketID()
	{
		return m_socketID;
	}

	void UserSecurityService::SetSocketID(int socketID)
	{
		m_socketID = socketID;
	}

	///////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	int UserSecurityService::AddCategory(int requestID, UserSecurityCategory *category)
	{
		vector<UserSecurityCategory> categories;
		categories.push_back(*category);
		int ret = Send(FUNCTIONID_USERSECURITY_ADDCATEGORIES, requestID, &categories);
		categories.clear();
		return ret > 0 ? 1 : 0;
	}

	int UserSecurityService::AddSecurities(int requestID, UserSecurityCategory *category)
	{
		vector<UserSecurityCategory> categories;
		categories.push_back(*category);
		int ret = Send(FUNCTIONID_USERSECURITY_ADDSECURITIES, requestID, &categories);
		categories.clear();
		return ret > 0 ? 1 : 0;
	}

	int UserSecurityService::DeleteCategory(int requestID, UserSecurityCategory *category)
	{
		vector<UserSecurityCategory> categories;
		categories.push_back(*category);
		int ret = Send(FUNCTIONID_USERSECURITY_DELETECATEGORIES, requestID, &categories);
		categories.clear();
		return ret > 0 ? 1 : 0;
	}

	int UserSecurityService::DeleteSecurities(int requestID, UserSecurityCategory *category)
	{
		vector<UserSecurityCategory> categories;
		categories.push_back(*category);
		int ret = Send(FUNCTIONID_USERSECURITY_DELETESECURITIES, requestID, &categories);
		categories.clear();
		return ret > 0 ? 1 : 0;
	}

	int UserSecurityService::GetCategories(int requestID, UserSecurityCategory *category)
	{
		vector<UserSecurityCategory> categories;
		categories.push_back(*category);
		int ret = Send(FUNCTIONID_USERSECURITY_GETCATEGORIES, requestID, &categories);
		categories.clear();
		return ret > 0 ? 1 : 0;
	}

	int UserSecurityService::GetCategories(vector<UserSecurityCategory> *categories, const char *body, int bodyLength)
	{
		Binary binary;
		binary.Write(body, bodyLength);
		int size = binary.ReadInt();
		for(int i = 0; i < size; i++)
		{
			UserSecurityCategory category;
			binary.ReadString(category.m_categoryID);
			category.m_userID = binary.ReadInt();
			binary.ReadString(category.m_name);
			binary.ReadString(category.m_codes);
			category.m_type = binary.ReadInt();
			category.m_orderNum = binary.ReadInt();
			categories->push_back(category);
		}
		return 1;
	}

	void UserSecurityService::OnReceive(CMessage *message)
	{
		BaseService::OnReceive(message);
		SendToListener(message);
	}

	int UserSecurityService::Send(int function, int requestID, vector<UserSecurityCategory> *categories)
	{
		int categoriesSize = (int)categories->size();
		int strSize = 0;
		for(int i = 0; i < categoriesSize; i++)
		{
			UserSecurityCategory category = (*categories)[i];
			strSize += sizeof(int) * 3 + CStrA::GetWStringCapacity(category.m_categoryID) + CStrA::GetWStringCapacity(category.m_codes) + CStrA::GetWStringCapacity(category.m_name) + 20;
		}
		Binary binary;
		binary.Initialize(strSize, false);
		binary.WriteInt(categoriesSize);
		for(int i = 0; i < categoriesSize; i++)
		{
			UserSecurityCategory category = (*categories)[i];
			binary.WriteString(category.m_categoryID);
			binary.WriteInt(category.m_userID);
			binary.WriteString(category.m_name);
			binary.WriteString(category.m_codes);
			binary.WriteInt(category.m_type);
			binary.WriteInt(category.m_orderNum);
		}
		char *ps = (char*)binary.GetBytes();
		CMessage message(GetGroupID(), GetServiceID(), function, GetSessionID(), requestID, m_socketID, 0, GetCompressType(), strSize, ps);
		return BaseService::Send(&message);
	}

	int UserSecurityService::UpdateCategory(int requestID, UserSecurityCategory *category)
	{
		vector<UserSecurityCategory> categories;
		categories.push_back(*category);
		int ret = Send(FUNCTIONID_USERSECURITY_UPDATECATEGORIES, requestID, &categories);
		categories.clear();
		return ret;
	}
}